ibc-core/ics24-host/types/src/identifiers/chain_id.rs

```diff
@@ -116,6 +115,173 @@ impl ChainId {
+                            }
+
+                            let next_value = map.next_value::<&str>()?;
+
+                            match parse_chain_id_string(next_value) {
+                                Ok((chain_name, rn)) => {
+                                    validate_prefix_length(chain_name, 1, 64).map_err(|_| {
+                                        Error::custom("invalid chain ID prefix length")
+                                    })?;
+
+                                    id = Some(next_value);
+                                    revision_number = Some(rn);
+                                }
+                                _ => {
+                                    validate_identifier_length(next_value, 1, 64).map_err(
+                                        |_| Error::custom("invalid chain ID identifier length"),
+                                    )?;
+
+                                    id = Some(next_value);
+                                    revision_number = Some(0);
+                                }
+                            }
+                        }
+                        Field::RevisionNumber => {
+                            let next_value = map.next_value::<&str>()?;
+                            let rev = u64::from_str(next_value).unwrap_or(0);
+
+                            if let Some(rn) = revision_number {
+                                if rev != 0 && rn != rev {
+                                    return Err(Error::custom(format_args!(
+                                        "chain ID revision numbers do not match; got `{}` and `{}`",
+                                        rn, rev,
+                                    )));
+                                }
+                            } else {
+                                revision_number = Some(rev);
+                            }
+                        }
+                    }
+                }
+
+                let id = id.ok_or_else(|| Error::missing_field("id"))?;
+
+                Ok(ChainId {
+                    id: id.into(),
+                    revision_number: revision_number.unwrap_or(0),
+                })
+            }
+        }
+
+        deserializer.deserialize_struct("ChainId", FIELDS, ChainIdVisitor)
+    }
+}
+
+#[cfg(feature = "borsh")]
+mod borsh_impls {
+    use borsh::maybestd::io::{self, Error, ErrorKind, Read};
+    use borsh::BorshDeserialize;
+
+    use super::*;
+
+    #[derive(BorshDeserialize)]
+    pub struct InnerChainId {
+        id: String,
+        revision_number: u64,
+    }
+
+    impl BorshDeserialize for ChainId {
+        fn deserialize_reader<R: Read>(reader: &mut R) -> io::Result<Self> {
+            let inner = InnerChainId::deserialize_reader(reader)?;
+
+            let Ok((_, rn)) = parse_chain_id_string(&inner.id) else {
+                return Err(Error::new(ErrorKind::Other, "failed to parse chain ID"));
+            };
```

Comment:
Despite not being parsed, a chain id string is not necessarily invalid. The `#[case(r#"{"id":"foo","revision_number":"0"}"#)]` should be successfully deserialized. 
Could you add tests for valid borsh desrialization cases as well?

Reply:
What about `{"id":"foo-bar-42","revision_number":"0"}`? The condition below accepts it but my intuition is that it should be rejected as well. IMO deserialised object should be the same as one constructed via `ChainId::from_str(&deserialised.id)`.

Reply:
It's possible to have `ChainId`s that don't strictly follow the format, but are still considered valid. `{"id":"foo-bar-42","revision_number":"0"}` is an example of a `ChainId` that is valid but that doesn't strictly follow the format.